torchvision/prototype/datapoints/_dataset_wrapper.py

```diff
@@ -0,0 +1,411 @@
+        self.transforms, dataset.transforms = dataset.transforms, None
+
+    def __getattr__(self, item):
+        with contextlib.suppress(AttributeError):
+            return object.__getattribute__(self, item)
+
+        return getattr(self._vision_dataset, item)
+
+    def __getitem__(self, idx):
+        # This gets us the raw sample since we disabled the transforms for the underlying dataset in the constructor
+        # of this class
+        sample = self._vision_dataset[idx]
+
+        sample = self._wrapper(self._vision_dataset, sample)
+
+        # Regardless of whether the user has supplied the transforms individually (`transform` and `target_transform`)
+        # or joint (`transforms`), we can access the full functionality through `transforms`
+        if self.transforms is not None:
+            sample = self.transforms(*sample)
+
+        return sample
+
+    def __len__(self):
+        return len(self._vision_dataset)
+
+
+def identity(item):
+    return item
+
+
+def list_of_dicts_to_dict_of_lists(list_of_dicts):
+    if not list_of_dicts:
+        return {}
+
+    dict_of_lists = {key: [value] for key, value in list_of_dicts[0].items()}
+    for dct in list_of_dicts[1:]:
+        for key, value in dct.items():
+            dict_of_lists[key].append(value)
+    return dict_of_lists
+
+
+def wrap_target_by_type(dataset, target, type_wrappers, *, fail_on=(), attr_name="target_type"):
+    if target is None:
+        return None
+
+    target_types = getattr(dataset, attr_name)
+
+    if any(target_type in fail_on for target_type in target_types):
+        raise RuntimeError(
+            f"{type(dataset).__name__} with target type(s) {sequence_to_str(fail_on, separate_last='or ')} "
+            f"is currently not supported by this wrapper. "
+            f"If this would be helpful for you, please open an issue at https://github.com/pytorch/vision/issues."
+        )
+
+    if not isinstance(target, (tuple, list)):
+        target = [target]
+
+    wrapped_target = tuple(
+        type_wrappers.get(target_type, identity)(item) for target_type, item in zip(target_types, target)
+    )
+
+    if len(wrapped_target) == 1:
+        wrapped_target = wrapped_target[0]
+
+    return wrapped_target
+
+
+@cache
+def get_categories(dataset):
+    categories_fn = {
+        datasets.Caltech256: lambda dataset: [name.rsplit(".", 1)[1] for name in dataset.categories],
+        datasets.CIFAR10: lambda dataset: dataset.classes,
+        datasets.CIFAR100: lambda dataset: dataset.classes,
+        datasets.FashionMNIST: lambda dataset: dataset.classes,
+        datasets.ImageNet: lambda dataset: [", ".join(names) for names in dataset.classes],
+        datasets.DatasetFolder: lambda dataset: dataset.classes,
+        datasets.ImageFolder: lambda dataset: dataset.classes,
+    }.get(type(dataset))
+    return categories_fn(dataset) if categories_fn is not None else None
+
+
+def classification_wrapper(dataset, sample):
```

Comment:
You are right that we only need to wrap the sample. However the canonical way to use transforms is by setting `dataset = datasets.ImageNet(..., transform=...)`. Since we need to wrap the raw data before it hits the transforms, we only have two options:

1. Wrap the dataset and monkeypatch `__getitem__` to inject our wrapping before we hit the transforms. 
2. Tell users to not use `transform`, `target_transform`, and `transforms` as before, and use them manually.

IMO, 2. is more disruptive to the workflow and thus I opted for 1. For completeness, if we are going for 2. after all, the whole chicken-and-egg problem goes away, since we then can instantiate the dataset before we instantiate the transform. Meaning, we could do something like

```python
dataset = datasets.ImageNet(...)

transform = transforms.Compose([
    V2WrapperTransform.from_vision_dataset(dataset),
    ...,
])

# showing the regular case of only using `transform`
# For other cases, the loop will have to be adjusted
for input, target in dataset:
    input = transform(input)
    ...
```

Reply:
I'm not sure I follow. If we didn't care about the `categories` meta-data, we wouldn't need to pass the dataset, correct?

Reply:
Nope. The wrapper needs to be adjusted depending on the dataset parameter. Take the `target_type` for example. Without knowing its exact value, we cannot provide a wrapper, since that information is missing from the sample. So `V2WrapperTransform.from_vision_dataset(...)` can either take the instantiated dataset which has this information, or we need to basically replicate the signature of the class, e.g.

```python
dataset = MyDataset(root, foo=2, bar="bar", baz=[-1.0, 2.0])
transform = V2WrapperTransform.from_dataset_args(MyDataset, foo=2, bar="bar", baz=[-1.0, 2.0])
```

Of course there are datasets for which the wrapper can be instantiated by just knowing the class, e.g. the classification datasets. But this doesn't hold in general.

Reply:
OK, I guess what I'm trying to say is: we only need to know the dataset instance *once*, not every single time `__getitem__()` gets called? I.e. we patch `__getitem__()` once and for all when `wrap_dataset_for_transforms_v2` is called and then the inner wrapper in `__getitem__()` should only need to take samples as input?

Reply:
Correct. I've opted to have a new object that provides the custom `__getitem__` over monkey-patching the original object, but I don't see a downside to the latter either.

Reply:
We don't have to monkey-patch the original dataset we just have to create the inner wrapper in `__init__()` of `_VisionDatasetDatapointWrapper`. It sounds unefficient to do all the checks in  `wrap_target_by_type()` when they just need to be done once at instanciation

Reply:
Fair enough. Let me find a better way to implement this.